### DISCUSSION TITLE
Is strimzi incompatible with K8s 1.21 because of service account token expiry?

### DISCUSSION THREAD

**Discussion by raflesw** [strimzi/strimzi-kafka-operator#6804] [Q&A]
After our K8s environment was upgraded to 1.21, we saw annotations like the following appear:

annotations.authentication.k8s.io/stale-token
subject: system:serviceaccount:eventbus:my-cluster-entity-operator, seconds after warning threshold: 3989

This is due to changes in token expiry in K8s 1.21 as described here:
https://docs.aws.amazon.com/eks/latest/userguide/troubleshooting.html#troubleshooting-boundservicetoken

It would appear that there is 90d grace period, after which tokens will be rejected.

It looks like the entity-operator needs to use a later client SDK version, or is there a workaround?

**Comment by scholzj** [comment]
What version of Strimzi you use? I haven't seen this annotation anywhere and I'm not aware of any issues with Kubernetes 1.21 or newer.

**Reply by raflesw** [reply]
Hi- thanks, we are using '0.26.1

**Reply by scholzj** [reply]
I guess there were some bumps to the Kubernetes client since then. But I'm not aware of any specific fixes around this. Also when searching for any existing issues, most of them seemed to be about som emore complicated connections from outside of the Kubernetes cluster and not about service accounts.

**Reply by balonik** [reply]
I see this using Strimzi `0.25` on k8s 1.21 cluster. This feature is in k8s since 1.21, see [release notes](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.21.md).

The 90d toke lifetime is enabled on EKS, vanilla k8s has standard 1 hour token lifetime.

Actually, AWS started to notify customers about it today, listing affected `ServiceAccounts`. In our case Strimzi `0.25` was among impacted ones. 

I suppose it depends what kind of k8s SDK is Strimzi using. If it is one of the official ones or some custom one. According to AWS docs using one of the official k8s SDK should resolve it:
Go v0.15.7 and later
Python v12.0.0 and later
Java v9.0.0 and later
JavaScript v0.10.3 and later
Ruby master branch
Haskell v0.3.0.0

**Reply by scholzj** [reply]
Strimzi - as most other Java apps I know - are using the Fabric8 Kubernetes client.

**Reply by scholzj** [reply]
> The 90d toke lifetime is enabled on EKS, vanilla k8s has standard 1 hour token lifetime.

This would suggest it should be even more visible on other clusters. But I do not see any issues in my Kube clusters (1.22, 1.23 and 1.24) even after 50+ days of runtime.

**Reply by balonik** [reply]
I found this on the Fabric8 https://github.com/fabric8io/kubernetes-client/issues/2271

**Reply by scholzj** [reply]
Hmm, I guess your search skills are better than mine. So that should give an answer. 

Still a bit puzzled by this -> Kube 1.21 is quite old. So I guess this is something what either nobody else enables or there has to be some unique AWS aspect to it.

**Reply by balonik** [reply]
It seems there is a "migration" period enabled by default in k8s that allows to use the token for 1 year even if it is "expired" already.
`These extended tokens would not expire and continue to be accepted within one year.` That's probably why there are no issues, yet.
This behavior is controlled by API server flag `--service-account-extend-token-expiration` which is `true` by default.

**Reply by balonik** [reply]
I suppose AWS just shorten the lifetime period from 1-year to 90d.

**Reply by tomsucho** [reply]
I also noticed that when you restart the pods, the api audit logs will stop having this annotation for an hour for that workload, and then it will start showing up again with the counter restarted from 0. 
```
annotations.authentication.k8s.io/stale-token 
subject: system:serviceaccount:<redacted namespace>:<redacted name>, seconds after warning threshold: 1850
```
So I think it can happen you may not see any problems if pods lifetime never goes beyond 90d (per AWS) or 1y (vanilla k8s)?

**Reply by balonik** [reply]
Yes, because if you restart Pod it would get the refreshed token mounted and use that.

**Reply by tomsucho** [reply]
Hey, so would that get fixed? At some point EKS will switch from 90d to the default 1h and then that would be an issue for sure, as we can't rely on pods restarting every 1h? Not even sure we can rely they would get restarted before every 90d tbh.

**Reply by scholzj** [reply]
The discussion above suggests that it is missing feature in the Fabric8 Kubernetes client. So it needs to be fixed there first. Once it is fixed there, we can update Strimzi to use a new version of it.

**Reply by tusharbhasme** [reply]
Can we switch to using the official Java client for k8s instead of using fabric8: https://github.com/kubernetes-client/java

**Reply by iciclespider** [reply]
This is because the Fabric8IO Java Client does not properly implement refreshing the service account token. See: 
 https://github.com/fabric8io/kubernetes-client/issues/2112#issuecomment-1128390799

**Reply by tomsucho** [reply]
btw. would you be able to tell, if that token gets stale i.e. after 90d in EKS would the containers malfunction in way that the liveness probes would catch it and automatically restart? or is it better to have a CronJob to do rolling restart of pods once a month to be on safe side?

**Reply by iciclespider** [reply]
> btw. would you be able to tell, if that token gets stale i.e. after 90d in EKS would the containers malfunction in way that the liveness probes would catch it and automatically restart? or is it better to have a CronJob to do rolling restart of pods once a month to be on safe side?

Neither. The 90d expiration is during the current grace period of transitioning to enforcing the 1 hour token expiration. IOW, eventually the tokens will expire after 1 hour.

**Reply by tomsucho** [reply]
@iciclespider the default 1h timer will happen in some future k8s version I guess. I'm asking what to do for 1.21 case or 1.22 on EKS, which both have expiration set to 90d, after which api server requests will get 40x (unless pods restarted in meantime as then token would recreate and have fresh 90d expiration again).

**Reply by lkoniecz** [reply]
@scholzj fabric has this one fixed https://github.com/fabric8io/kubernetes-client/pull/3445

**Reply by scholzj** [reply]
@lkoniecz Strimzi uses Fabric8 5.8.0 or newer since 0.26.0 (done in PR #5613). So if this fixed it, it would explain seeing it with Strimzi 0.25 (reported by @balonik). But not with Strimzi 0.26.1 (mentioned by @raflesw).

**Reply by psemeniuk** [reply]
@scholzj because fabric8 maintainers decided to not to follow official guides, but instead they implemented fallback on 401. Details here https://github.com/fabric8io/kubernetes-client/pull/3445#issuecomment-922221613

So finally you will be still getting stale-token warnings, and there is no way to test it without container uptime longer than grace period (90d in EKS / one year in bare k8s).

**Reply by tyagian** [reply]
What should be the latest version to have this fix? Strimzi 0.25 ?
Any version after Strimzi 0.25 will be getting stale-token warnings?
It seems fabric updated in 0.27.1 also.
https://github.com/strimzi/strimzi-kafka-operator/releases/tag/0.27.1

But I see warnings in 0.27.1 also. Any workaround yet?

**Reply by chadlangston** [reply]
anyone know the answer to @tyagian's question?

thanks!

**Reply by scholzj** [reply]
I think everything is said above. I'm not aware of any issues with any supported Kubernetes version.

**Reply by chadlangston** [reply]
cool. so, am i reading the above correctly that the token refresh issue should be fixed in the strimzi kafka operator v0.26 or higher, but that we may still see stale-token warnings due to the method that fabric8 is using to refresh the token?

you seem to suggest that this fix should be present in 0.31 or higher here
https://github.com/strimzi/strimzi-kafka-operator/issues/6845

am i confusing two different issues?
